test/unit/DominantJuice.unit.t.sol

```diff
@@ -0,0 +1,962 @@
+        vm.mockCall(
+            address(directory),
+            abi.encodeCall(directory.isTerminalOf, (projectID, ethPaymentTerminal)),
+            abi.encode(true)
+        );
+
+        vm.prank(address(ethPaymentTerminal)); // Payment terminal calling
+        vm.expectRevert(abi.encodeWithSelector(DominantJuice.IncorrectProjectID.selector));
+        dominantJuice.didRedeem(didRedeemData);
+    }
+
+    // PASSING
+    // Test in case Juicebox architecture goes wonky.
+    function test_didRedeem_revertsIfNotPledger() public bonusDeposited {
+        failedCycleHasExpired();
+        vm.warp(cycleExpiryDate + 100);
+
+        didRedeemData.holder = rando;
+        didRedeemData.projectId = projectID;
+
+        vm.mockCall(
+            address(directory),
+            abi.encodeCall(directory.isTerminalOf, (projectID, ethPaymentTerminal)),
+            abi.encode(true)
+        );
+
+        vm.prank(address(ethPaymentTerminal));
+        vm.expectRevert(abi.encodeWithSelector(DominantJuice.MustBePledger.selector));
+        dominantJuice.didRedeem(didRedeemData);
+    }
+
+    // PASSING
+    // Test in case Juicebox architecture goes wonky.
+    function test_didRedeem_revertsIfAlreadyWithdrawn() public bonusDeposited {
+        failedCycleHasExpired();
+        didRedeemData.holder = pledger1;
+        didRedeemData.projectId = projectID;
+
+        // Pledger1 redeems
+        redeem(pledger1);
+
+        vm.mockCall(
+            address(directory),
+            abi.encodeCall(directory.isTerminalOf, (projectID, ethPaymentTerminal)),
+            abi.encode(true)
+        );
+
+        // Pledger1 tries to redeem again via JB.
+        vm.prank(address(ethPaymentTerminal));
+        vm.expectRevert(DominantJuice.AlreadyWithdrawnRefund.selector);
+        dominantJuice.didRedeem(didRedeemData);
+    }
+
+    // PASSING
+    function testFuzz_didRedeem_happyPath(uint256 _pledgeAmount) public bonusDeposited {
+        // Pledger1 pledges up to and excluding cycleTarget amount and time advances past expiry date.
+        _pledgeAmount = bound(_pledgeAmount, MIN_PLEDGE_AMOUNT, CYCLE_TARGET - 1);
+        pledge(pledger1, _pledgeAmount);
+        vm.warp(cycleExpiryDate + 100);
+
+        didRedeemData.holder = pledger1;
+        didRedeemData.projectId = projectID;
+
+        vm.mockCall(
+            address(directory),
+            abi.encodeCall(directory.isTerminalOf, (projectID, ethPaymentTerminal)),
+            abi.encode(true)
+        );
+        vm.expectCall(address(directory), abi.encodeCall(directory.isTerminalOf, (projectID, ethPaymentTerminal)));
+
+        (uint256 pledgerWeight, uint256 totalWeight) = dominantJuice.getPledgerAndTotalWeights(pledger1);
+        DominantJuice.Campaign memory campaign = dominantJuice.getCampaignInfo();
+
+        uint256 pledgerRefundBonus = (pledgerWeight / totalWeight) * campaign.totalRefundBonus;
+
+        if (dominantJuice.getBalance() < pledgerRefundBonus) {
```

Comment:
whenever I see an if block in a test, something is wrong. The problem here is that you get different results. You should have two different tests and each one has a guarantee of the outcome

Reply:
For this comment and the next three involving ifs: So again, the guy I’m learning under right now incorporates if statements in his unit tests and has encouraged me to do so. He seems to be a Forge expert, so maybe it’s just with Forge?... BUT, again, I want to make sure that I do right by you with this, so I can remove them for you.

Reply:
Done.